dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/clients/common/UserInfoHelper.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * IBM Confidential
+ *
+ * OCO Source Materials
+ *
+ * Copyright IBM Corp. 2018
+ *
+ * The source code for this program is not published or otherwise divested
+ * of its trade secrets, irrespective of what has been deposited with the
+ * U.S. Copyright Office.
+ */
+package com.ibm.ws.security.openidconnect.clients.common;
+
+import java.util.Map;
+
+import javax.net.ssl.SSLSocketFactory;
+
+import org.apache.http.HttpResponse;
+import org.apache.http.util.EntityUtils;
+
+import com.ibm.json.java.JSONObject;
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.security.openidconnect.common.Constants;
+import com.ibm.ws.webcontainer.security.ProviderAuthenticationResult;
+
+/**
+ * Utility methods to retrieve UserInfo data, validate it, and update the Subject with it.
+ */
+public class UserInfoHelper {
+    private static final TraceComponent tc = Tr.register(UserInfoHelper.class, TraceConstants.TRACE_GROUP, TraceConstants.MESSAGE_BUNDLE);
+    private ConvergedClientConfig clientConfig = null;
+
+    public UserInfoHelper(ConvergedClientConfig config) {
+        this.clientConfig = config;
+    }
+
+    public boolean willRetrieveUserInfo() {
+        return (clientConfig.getUserInfoEndpointUrl() != null && clientConfig.isUserInfoEnabled() == true);
+    }
+
+    /**
+     * get userinfo from provider's UserInfo Endpoint if configured and active.
+     * If successful, update properties in the ProviderAuthenticationResult
+     *
+     * @return true if PAR was updated with userInfo
+     *
+     */
+    public boolean getUserInfo(ProviderAuthenticationResult oidcResult,
+            SSLSocketFactory sslSocketFactory, String accessToken, String subjectFromIdToken) {
+
+        if (!willRetrieveUserInfo() || accessToken == null) {
+            return false;
+        }
+        String userInfoStr = getUserInfoFromURL(clientConfig, sslSocketFactory, accessToken);
+
+        if (userInfoStr == null) {
+            return false;
+        }
+
+        if (!isUserInfoValid(userInfoStr, subjectFromIdToken)) {
+            return false;
+        }
+
+        updateAuthenticationResultPropertiesWithUserInfo(oidcResult, userInfoStr);
+        return true;
+    }
+
+    protected void updateAuthenticationResultPropertiesWithUserInfo(ProviderAuthenticationResult oidcResult, String userInfoStr) {
+        oidcResult.getCustomProperties().put(Constants.USERINFO_STR, userInfoStr);
+    }
+
+    // per oidc-connect-core-1.0 sec 5.3.2, sub claim of userinfo response must match sub claim in id token.
+    protected boolean isUserInfoValid(String userInfoStr, String subClaim) {
+        String userInfoSubClaim = getUserInfoSubClaim(userInfoStr);
+        if (userInfoSubClaim == null || subClaim == null || userInfoSubClaim.compareTo(subClaim) != 0) {
```

Comment:
Are there benefits to using `String.compareTo()` as opposed to `String.equals()`?